refactor: deduplicate integer serialization in RollingBloom benchmark

phyBrackets opened the pull request:
<!--
*** Please remove the following help text before submitting: ***

Pull requests without a rationale and clear improvement may be closed
immediately.

GUI-related pull requests should be opened against
https://github.com/bitcoin-core/gui
first. See CONTRIBUTING.md
-->

<!--
Please provide clear motivation for your patch and explain how it improves
Bitcoin Core user experience or Bitcoin Core developer experience
significantly:

* Any test improvements or new tests that improve coverage are always welcome.
* All other changes should have accompanying unit tests (see `src/test/`) or
  functional tests (see `test/`). Contributors should note which tests cover
  modified code. If no tests exist for a region of modified code, new tests
  should accompany the change.
* Bug fixes are most welcome when they come with steps to reproduce or an
  explanation of the potential issue as well as reasoning for the way the bug
  was fixed.
* Features are welcome, but might be rejected due to design or scope issues.
  If a feature is based on a lot of dependencies, contributors should first
  consider building the system outside of Bitcoin Core, if possible.
* Refactoring changes are only accepted if they are required for a feature or
  bug fix or otherwise improve developer experience significantly. For example,
  most "code style" refactoring changes require a thorough explanation why they
  are useful, what downsides they have and why they *significantly* improve
  developer experience or avoid serious programming bugs. Note that code style
  is often a subjective matter. Unless they are explicitly mentioned to be
  preferred in the [developer notes](/doc/developer-notes.md), stylistic code
  changes are usually rejected.
-->

<!--
Bitcoin Core has a thorough review process and even the most trivial change
needs to pass a lot of eyes and requires non-zero or even substantial time
effort to review. There is a huge lack of active reviewers on the project, so
patches often sit for a long time.
--> 
refactor: deduplicate integer serialization in RollingBloom benchmark and i think we have to follow don't repeat yourself pattern here! I don't think that this much affect the memory and efficiency because it is still in constant time . But surely it reduces some readability of code which is not the big headache for sure.

theStack reviewed (commented):
Seems like you could also simply use the helpers `WriteLE32` and `WriteBE32` from the `<crypto/common.h>` header in this case to avoid manual bit-fiddling:
https://github.com/bitcoin/bitcoin/blob/d0bf9bb6a539f151ec92725d20a2b6c22cb095a5/src/crypto/common.h#L44-L48

https://github.com/bitcoin/bitcoin/blob/d0bf9bb6a539f151ec92725d20a2b6c22cb095a5/src/crypto/common.h#L77-L81

As PR and commit subject I'd suggest using something more concrete like "refactor: deduplicate integer serialization in RollingBloom benchmark", rather than a (rhetorical) question.

phyBrackets commented:
> Seems like you could also simply use the helpers `WriteLE32` and `WriteBE32` from the `<crypto/common.h>` header in this case to avoid manual bit-fiddling:
> 
> https://github.com/bitcoin/bitcoin/blob/d0bf9bb6a539f151ec92725d20a2b6c22cb095a5/src/crypto/common.h#L44-L48

> https://github.com/bitcoin/bitcoin/blob/d0bf9bb6a539f151ec92725d20a2b6c22cb095a5/src/crypto/common.h#L77-L81
> 
> As PR and commit subject I'd suggest using something more concrete like "refactor: deduplicate integer serialization in RollingBloom benchmark", rather than a (rhetorical) question.

phyBrackets commented:
Thanks for the suggestion. I update the following.
@theStack

phyBrackets commented:
would you tell me @theStack why is there some checks were not successful? and how can i resolve it?

sipa commented on src/bench/rollingbloom.cpp:
```diff
@@ -13,16 +13,13 @@ static void RollingBloom(benchmark::Bench& bench)
-        data[1] = (count >> 8) & 0xFF;
-        data[2] = (count >> 16) & 0xFF;
-        data[3] = (count >> 24) & 0xFF;
+        for(int i = 0; i <= 3; i++){
```
This is just `WriteLE32(data, count);`.

sipa commented on src/bench/rollingbloom.cpp:
```diff
@@ -13,16 +13,13 @@ static void RollingBloom(benchmark::Bench& bench)
-        data[1] = (count >> 16) & 0xFF;
-        data[2] = (count >> 8) & 0xFF;
-        data[3] = count & 0xFF;
+        for(int i = 3; i >= 0; i--){
```
This is just `WriteBE32(data, count);`.

phyBrackets replied:
so, should i change the following to just a function call by adding `common.h` header ?

sipa replied:
Yeah, better than duplicating the logic here/

fanquake commented:
Once you've addressed the suggestions, make sure you squash your changes to a single commit, with a proper commit message.

phyBrackets commented:
> Once you've addressed the suggestions, make sure you squash your changes to a single commit, with a proper commit message.


@fanquake will take care of that. Thanks.

phyBrackets commented:
> Seems like you could also simply use the helpers `WriteLE32` and `WriteBE32` from the `<crypto/common.h>` header in this case to avoid manual bit-fiddling:
> 
> https://github.com/bitcoin/bitcoin/blob/d0bf9bb6a539f151ec92725d20a2b6c22cb095a5/src/crypto/common.h#L44-L48
> 
> https://github.com/bitcoin/bitcoin/blob/d0bf9bb6a539f151ec92725d20a2b6c22cb095a5/src/crypto/common.h#L77-L81
> 
> As PR and commit subject I'd suggest using something more concrete like "refactor: deduplicate integer serialization in RollingBloom benchmark", rather than a (rhetorical) question.


Isn't this bit inefficient because we are actually dropping the whole content of  `common.h` ?

phyBrackets commented:
> > Seems like you could also simply use the helpers `WriteLE32` and `WriteBE32` from the `<crypto/common.h>` header in this case to avoid manual bit-fiddling:
> > https://github.com/bitcoin/bitcoin/blob/d0bf9bb6a539f151ec92725d20a2b6c22cb095a5/src/crypto/common.h#L44-L48
> > 
> > https://github.com/bitcoin/bitcoin/blob/d0bf9bb6a539f151ec92725d20a2b6c22cb095a5/src/crypto/common.h#L77-L81
> > 
> > As PR and commit subject I'd suggest using something more concrete like "refactor: deduplicate integer serialization in RollingBloom benchmark", rather than a (rhetorical) question.
> 
> Isn't this bit inefficient because we are actually dropping the whole content of `common.h` ?


Hey @sipa what's your views on this before making a function call ?

sipa commented:
I don't understand your question. Using ReadLE32/WriteLE32 is almost certainly more efficient than reimplementing it yourself (they're written in a way that the compiler can easily optimize out). Not that ~nanosecond level performance is relevant here...

phyBrackets commented:
oh , okay ! Then it's fine I think.

phyBrackets commented:
Sorry for the direct commit , need to do some more changes.

phyBrackets commented:
Hey, any update on this?

MarcoFalke commented:
I haven't looked at the changes, but this can't be merged unless squashed.

Please squash your commits according to https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#squashing-commits

phyBrackets commented:
Cool, will do that soon. Thanks!

phyBrackets commented:
Hey, I'm having some problem with squashing the commits. Should I make new PR with final changes? Will close this and refer to this PR! If you would allow?

fanquake commented:
>  Should I make new PR with final changes?

No. Please don't open new PRs for the same changes. You might need to take another look through: https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#squashing-commits/.

phyBrackets commented:
Hey @fanquake , I tried to rebase/squash , I have some problems with branches I think , there is branch conflict . I really don't want to open a new PR but it seems the only option , by the way if you would allow, I'll open that and set the reference to this PR , and will close this pr ?

sipa commented:
@phyBrackets It's probably worthwhile learning how to resolve these issues anyway, if you want to continue contributing, whether you do that for this PR or a future one.

MarcoFalke commented on src/bench/rollingbloom.cpp:
```diff
@@ -5,24 +5,20 @@
 {
     CRollingBloomFilter filter(120000, 0.000001);
     std::vector<unsigned char> data(32);
+    unsigned char* dataPtr = &data[0];
```
```suggestion
    unsigned char* dataPtr = data.data();
```

Shouldn't this use the C++11 member function?

phyBrackets replied:
Yeah, resolve this with the new commit.

phyBrackets commented:
> @phyBrackets It's probably worthwhile learning how to resolve these issues anyway, if you want to continue contributing, whether you do that for this PR or a future one.

Hey @sipa , I'm really trying and searching around how to resolve this issue but I'm not able to , actually when I' trying to run `git rebase -i HEAD~7` , this only shows the first 4 commits.

MarcoFalke commented:
Your branch contains merge commits, which can't be squashed in a rebase.

Something like this might work (untested!!):

```
git checkout patch-1                                       # Verify this is on commit aa91fd400a6a23feeacdff0d2adacd18dd313e47
git fetch origin 5d254a234d8c1569b0161264cc6d5d8d0ce0d864  # Fetch the latest master
git merge 5d254a234d8c1569b0161264cc6d5d8d0ce0d864         # Merge the latest master
git reset --soft 5d254a234d8c1569b0161264cc6d5d8d0ce0d864  # Discard commit history, but keep all changes staged
git commit --message="bla bla"                             # Create a commit with all staged changes

phyBrackets commented:
I tried this , it seems more mess tho.

MarcoFalke commented:
After the steps in https://github.com/bitcoin/bitcoin/pull/24088#issuecomment-1045967132 you need to force push, not merge

phyBrackets committed:
Remove the deduplicating of data in rollingbloom.cpp

phyBrackets commented:
Thank you sm @MarcoFalke . It seems like it works fine.

theStack reviewed (approved):
Code-review ACK e1fcf8dc148615842341726c6aa35bb00c35f1d0

Nit: I don't think the `dataPtr` pointer is really needed here, you could just call the serialization helper functions with `data.data()` as first parameter each.

phyBrackets commented:
Cool, but I think it is more readable through this . But yeah if this affects on unnecessary memory use. We can remove that. Let me know your further take on this.

laanwj commented:
> Nit: I don't think the dataPtr pointer is really needed here, you could just call the serialization helper functions with data.data() as first parameter each.

+1. The assignment is unnecessary. I doubt it affects memory use in any significant way but I see no reason not to write it shorter if you're touching this code anyway.

fanquake commented:
I've just gone ahead and fixed this up in #24784, given this is straightforward, and likely been open long enough.
